ports/stm32/powerctrlboot.c

```diff
@@ -460,13 +460,68 @@ void SystemClock_Config(void) {
 #include "stm32wlxx_ll_utils.h"
 
 void SystemClock_Config(void) {
-    // Set flash latency
+    // Set flash latency (2 wait states, sysclk > 36MHz)
     LL_FLASH_SetLatency(LL_FLASH_LATENCY_2);
     while (LL_FLASH_GetLatency() != LL_FLASH_LATENCY_2) {
     }
 
     LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
 
+    #if MICROPY_HW_CLK_USE_HSE
+    // Enable the 32MHz external oscillator and 48MHZ SYSCLK via PLL
+
+    #if MICROPY_HW_CLK_USE_BYPASS
+    // Use "bypass power" option, port PB0_VDDTCXO supplies TCXO
+    // (STM32WL5x has no other HSE bypass mode.)
+
+    // "PB0 must be configured in analog mode prior enabling the HSE"
+    //
+    // Note: PB0 analog mode muxes PB0_VDDTCXO pin to the VDDTCXO regulator, set
+    // to default voltage of 1.7V.  Changing this voltage requires initializing
+    // the SUBGHZ radio and sending a Set_Tcxo command to it.
+    //
+    // For the Nucelo-WL55 board, ST uses the NDK "NT2016SF-32M-END5875A" TCXO
+    // which has no publicly available datasheet. However, the ST code for this
+    // board always keeps the pin at the default 1.7V voltage level so changing
+    // the level would only be needed if a different TCXO is used.
+    //
+    // (Note also that setting pin PB0 as a push-pull GPIO output is technically
+    // possible too, but 3.3V will be too high for many TCXOs.)
+    GPIO_InitTypeDef GPIO_InitStructure = {
+        .Pin = GPIO_PIN_0,
+        .Mode = GPIO_MODE_ANALOG,
+        .Pull = GPIO_NOPULL,
+    };
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);
+
+    #endif // MICROPY_HW_CLK_USE_BYPASS
+
+    __HAL_RCC_HSE_CONFIG(MICROPY_HW_RCC_HSE_STATE);
+    while (!__HAL_RCC_GET_FLAG(RCC_FLAG_HSERDY)) {
+        // Wait for HSE Ready signal
+    }
+
+    // Configure the PLL
+    __HAL_RCC_PLL_CONFIG(
+        RCC_PLLSOURCE_HSE, // 32MHz
+        RCC_PLLM_DIV2,     // VCO input 16MHz (recommended in ST docs)
+        6,                 // f_N = 6*16MHz = 96MHz
+        RCC_PLLP_DIV2,     // f_P = 48Mhz
+        RCC_PLLQ_DIV2,     // f_Q = 48MHz
+        RCC_PLLR_DIV2);    // f_R = 48MHz
+    __HAL_RCC_PLL_ENABLE();
+    __HAL_RCC_PLLCLKOUT_ENABLE(RCC_PLL_SYSCLK);
+    while (!__HAL_RCC_GET_FLAG(RCC_FLAG_PLLRDY)) {
+        // Wait for PLL to lock
+    }
+
+    __HAL_RCC_SYSCLK_CONFIG(RCC_SYSCLKSOURCE_PLLCLK);
+    while (__HAL_RCC_GET_SYSCLK_SOURCE() != RCC_SYSCLKSOURCE_STATUS_PLLCLK) {
```

Comment:
Please use the `LL_RCC_xxx` functions instead of `__HAL_RCC_xxx` (see just below for examples).  The LL is much preferred over the old HAL.

Reply:
@dpgeorge I switched from LL to HAL here because there's some register bit sequencing in the ` __HAL_RCC_HSE_CONFIG` macro that seemed unnecessary to recreate here, similarly the PLL calculation macro. However, I'll switch it back.

(As I'm sure you realise, various parts of the stm32 port contain different configuration methods so it's hard to guess which one is the preferred level to integrate against.)

Reply:
Rewritten to use LL[*], confirmed system frequency and LoRa modem functions are still the same.

[*] Except PLL config which is raw register write as the only LL method seems to be inexplicably split into three overlapping register operations.

Reply:
But it should be easy to at least just copy the code below that sets the MSI as the clock source.  That will use the same number of lines and be consistent with how MSI is done.

For the PLL... it can stay using HAL if that's simpler (but I think there's also a special LL function to configure the PLL values).

Reply:
Looks like our comments crossed, but let me know if you want the PLL config lines converted back to HAL.